test/libsolidity/ABIEncoderTests.cpp

```diff
@@ -0,0 +1,255 @@
+		}
+	)";
+	compileAndRun(sourceCode);
+	callContractFunction("f()");
+	REQUIRE_LOG_DATA(encodeArgs(
+		string(3, 0) + string("\x0a"), string("\xf1\xf2"),
+		0xff, 0xff, u256(-1), u256(1)
+	));
+}
+
+BOOST_AUTO_TEST_CASE(storage_byte_array)
+{
+	char const* sourceCode = R"(
+		contract C {
+			bytes short;
+			bytes long;
+			event E(bytes s, bytes l);
+			function f() {
+				short = "123456789012345678901234567890a";
+				long = "ffff123456789012345678901234567890afffffffff123456789012345678901234567890a";
+				E(short, long);
+			}
+		}
+	)";
+	compileAndRun(sourceCode);
+	callContractFunction("f()");
+	REQUIRE_LOG_DATA(encodeArgs(
+		0x40, 0x80,
+		31, string("123456789012345678901234567890a"),
+		75, string("ffff123456789012345678901234567890afffffffff123456789012345678901234567890a")
+	));
+}
+
+BOOST_AUTO_TEST_CASE(storage_array)
+{
+	char const* sourceCode = R"(
+		contract C {
+			address[3] addr;
+			event E(address[3] a);
+			function f() {
+				assembly {
+					sstore(0, sub(0, 1))
+					sstore(1, sub(0, 2))
+					sstore(2, sub(0, 3))
+				}
+				E(addr);
+			}
+		}
+	)";
+	compileAndRun(sourceCode);
+	callContractFunction("f()");
+	REQUIRE_LOG_DATA(encodeArgs(u160(-1), u160(-2), u160(-3)));
+}
+
+BOOST_AUTO_TEST_CASE(storage_array_dyn)
+{
+	char const* sourceCode = R"(
+		contract C {
+			address[] addr;
+			event E(address[] a);
+			function f() {
+				addr.push(1);
+				addr.push(2);
+				addr.push(3);
+				E(addr);
+			}
+		}
+	)";
+	compileAndRun(sourceCode);
+	callContractFunction("f()");
+	REQUIRE_LOG_DATA(encodeArgs(0x20, 3, u160(1), u160(2), u160(3)));
+}
+
+BOOST_AUTO_TEST_CASE(storage_array_compact)
+{
+	char const* sourceCode = R"(
+		contract C {
+			int72[] x;
+			event E(int72[]);
+			function f() {
+				x.push(-1);
+				x.push(2);
+				x.push(-3);
+				x.push(4);
+				x.push(-5);
+				x.push(6);
+				x.push(-7);
+				x.push(8);
+				E(x);
+			}
+		}
+	)";
+	compileAndRun(sourceCode);
+	callContractFunction("f()");
+	REQUIRE_LOG_DATA(encodeArgs(
+		0x20, 8, u256(-1), 2, u256(-3), 4, u256(-5), 6, u256(-7), 8
+	));
+}
+
+BOOST_AUTO_TEST_CASE(external_function)
+{
+	char const* sourceCode = R"(
+		contract C {
+			event E(function(uint) external returns (uint), function(uint) external returns (uint));
+			function(uint) external returns (uint) g;
+			function f(uint) returns (uint) {
+				g = this.f;
+				E(this.f, g);
+			}
+		}
+	)";
+	compileAndRun(sourceCode);
+}
+
+BOOST_AUTO_TEST_CASE(external_function_cleanup)
+{
+	char const* sourceCode = R"(
+		contract C {
+			event E(function(uint) external returns (uint), function(uint) external returns (uint));
+			function(uint) external returns (uint) g;
+			function f(uint) returns (uint) {
+				function(uint) external returns (uint)[1] memory h;
+				assembly { sstore(0, sub(0, 1)) mstore(h, sub(0, 1)) }
+				E(h[0], g);
+			}
+		}
+	)";
+	compileAndRun(sourceCode);
+	callContractFunction("f(uint256)");
+	REQUIRE_LOG_DATA(encodeArgs(string(24, 0xff), string(24, 0xff)));
```

Comment:
C:\projects\solidity\test\libsolidity\ABIEncoderTests.cpp(248): error C2220: warning treated as error - no 'object' file generated [C:\projects\solidity\build\test\soltest.vcxproj]
C:\projects\solidity\test\libsolidity\ABIEncoderTests.cpp(248): warning C4309: 'argument': truncation of constant value [C:\projects\solidity\build\test\soltest.vcxproj]